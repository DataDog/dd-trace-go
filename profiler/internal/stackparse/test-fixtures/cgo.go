//+build ignore

// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2016 Datadog, Inc.

package main

import (
	"runtime/pprof"
	/*
		#include <unistd.h>

		extern int goSleep(int usec);

		static void __attribute__ ((noinline)) c_sleep(int usec) {
			usleep(usec);
		}

		static void c_sleep_loop(int usec) {
			while (1) {
				c_sleep(usec);
			}
		}

		static void c_go_sleep_loop(int usec) {
			while (1) {
				goSleep(usec);
			}
		}
	*/
	"C"
)
import (
	"os"
	"time"
)

func main() {
	go cSleepLoop(time.Second)
	go cGoSleepLoop(time.Second)

	time.Sleep(time.Second / 10)
	pprof.Lookup("goroutine").WriteTo(os.Stdout, 2)
}

func cSleepLoop(d time.Duration) {
	C.c_sleep_loop(C.int(d.Microseconds()))
}

func cGoSleepLoop(d time.Duration) {
	C.c_go_sleep_loop(C.int(d.Microseconds()))
}

//export goSleep
func goSleep(usec C.int) C.int {
	time.Sleep(time.Duration(usec) * time.Microsecond)
	return 0
}
