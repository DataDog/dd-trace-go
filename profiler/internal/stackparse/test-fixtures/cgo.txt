goroutine 1 [running]:
runtime/pprof.writeGoroutineStacks(0x4124060, 0xc000010018, 0x1000000000030, 0x4713e40)
	/usr/local/Cellar/go/1.15.6/libexec/src/runtime/pprof/pprof.go:693 +0x9f
runtime/pprof.writeGoroutine(0x4124060, 0xc000010018, 0x2, 0x0, 0x50)
	/usr/local/Cellar/go/1.15.6/libexec/src/runtime/pprof/pprof.go:682 +0x45
runtime/pprof.(*Profile).WriteTo(0x41b45a0, 0x4124060, 0xc000010018, 0x2, 0xffffffff, 0xc00007a058)
	/usr/local/Cellar/go/1.15.6/libexec/src/runtime/pprof/pprof.go:331 +0x3f2
main.main()
	/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go:40 +0xbb

goroutine 6 [syscall]:
main._Cfunc_c_sleep_loop(0xf4240)
	_cgo_gotypes.go:56 +0x45
main.cSleepLoop(0x3b9aca00)
	/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go:44 +0x45
created by main.main
	/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go:36 +0x45

goroutine 7 [sleep, locked to thread]:
time.Sleep(0x3b9aca00)
	/usr/local/Cellar/go/1.15.6/libexec/src/runtime/time.go:188 +0xbf
main.goSleep(...)
	/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go:53
main._cgoexpwrap_65ad5df3a112_goSleep(0xf4240, 0xc000001b00)
	_cgo_gotypes.go:73 +0x34
main._Cfunc_c_go_sleep_loop(0xf4240)
	_cgo_gotypes.go:43 +0x45
main.cGoSleepLoop(0x3b9aca00)
	/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go:48 +0x45
created by main.main
	/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go:37 +0x66
