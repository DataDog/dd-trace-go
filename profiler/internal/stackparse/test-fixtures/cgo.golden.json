{
  "Errors": null,
  "Goroutines": [
    {
      "ID": 1,
      "State": "running",
      "Wait": 0,
      "LockedToThread": false,
      "Stack": [
        {
          "Func": "runtime/pprof.writeGoroutineStacks",
          "File": "/usr/local/Cellar/go/1.15.6/libexec/src/runtime/pprof/pprof.go",
          "Line": 693
        },
        {
          "Func": "runtime/pprof.writeGoroutine",
          "File": "/usr/local/Cellar/go/1.15.6/libexec/src/runtime/pprof/pprof.go",
          "Line": 682
        },
        {
          "Func": "runtime/pprof.(*Profile).WriteTo",
          "File": "/usr/local/Cellar/go/1.15.6/libexec/src/runtime/pprof/pprof.go",
          "Line": 331
        },
        {
          "Func": "main.main",
          "File": "/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go",
          "Line": 40
        }
      ],
      "FramesElided": false,
      "CreatedBy": null
    },
    {
      "ID": 6,
      "State": "syscall",
      "Wait": 0,
      "LockedToThread": false,
      "Stack": [
        {
          "Func": "main._Cfunc_c_sleep_loop",
          "File": "_cgo_gotypes.go",
          "Line": 56
        },
        {
          "Func": "main.cSleepLoop",
          "File": "/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go",
          "Line": 44
        }
      ],
      "FramesElided": false,
      "CreatedBy": {
        "Func": "main.main",
        "File": "/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go",
        "Line": 36
      }
    },
    {
      "ID": 7,
      "State": "sleep",
      "Wait": 0,
      "LockedToThread": true,
      "Stack": [
        {
          "Func": "time.Sleep",
          "File": "/usr/local/Cellar/go/1.15.6/libexec/src/runtime/time.go",
          "Line": 188
        },
        {
          "Func": "main.goSleep",
          "File": "/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go",
          "Line": 53
        },
        {
          "Func": "main._cgoexpwrap_65ad5df3a112_goSleep",
          "File": "_cgo_gotypes.go",
          "Line": 73
        },
        {
          "Func": "main._Cfunc_c_go_sleep_loop",
          "File": "_cgo_gotypes.go",
          "Line": 43
        },
        {
          "Func": "main.cGoSleepLoop",
          "File": "/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go",
          "Line": 48
        }
      ],
      "FramesElided": false,
      "CreatedBy": {
        "Func": "main.main",
        "File": "/Users/felix.geisendoerfer/go/src/github.com/DataDog/dd-trace-go/profiler/internal/stackparse/test-fixtures/cgo.go",
        "Line": 37
      }
    }
  ]
}
